### DISCUSSION TITLE
Is it possible to create a new folder, or does move do this automatically?

### DISCUSSION THREAD

**Discussion by mttjhn** [SilentVoid13/Templater#333] [Help] [answered]
I'm fascinated with the idea of automatic filing of notes, and I'd love to create a new sub-folder if it doesn't exist (representing a project or issue where I'll file notes related to that). Is creating a sub-folder something that is possible to do using Templater? If so, how would you check first if the folder exists in your code? Or am I totally wrong and does Templater do that automatically when you move a file to a path that doesn't exist?

**Comment by kquinsland** [comment]
I am a +1 for this.  I love the move function and am starting to make most of my notes 'automatically file'... when the destination directory exists.

I was thinking about creating a function that would run `mkdir -p` or equiv but came here to check if `move()` could do this automatically or not.

**Comment by welpdx** [comment, marked as answer]
Hey guys, 

If you need to create a new folder, why not try the `createFolder()` function in the Obsidian Api? ([It is referenced on line 1004](https://github.com/obsidianmd/obsidian-api/blob/bde556afa033e909ebfb9fcee8f5ef288276f78f/obsidian.d.ts#L3004))

You can check if path exists. If it doesn't, do:
```
await this.app.vault.createFolder("/foo/bar/")
```

![gif](https://i.imgur.com/ywJF5hx.gif)

Hope that helps,

Welp~

Edit: <sub>  something o.o </sub>

**Reply by mttjhn** [reply]
Awesome! I was unaware you could do this.  Essentially, I'm doing the following in my template now:
```
let taskFolderPath = tp.file.folder(true) + "/" + taskNumber + "/";
if (!tp.file.exists(taskFolderPath)) {
    await this.app.vault.createFolder(taskFolderPath)
}
await tp.file.move(taskFolderPath + titleName);
```

**Reply by welpdx** [reply]
Glad to be of help! I like the script you made too!

Welp

**Reply by Smitty010** [reply]
I have been playing with this snippet (as above) and I don't think it works in all cases.  Here's the problem.  this.app.vault.createFolder expects a folder path relative to your vault folder.  tp.file.move also expects a folder path relative to your vault folder.  However, tp.file.exists expects a folder path relative to the **current folder**.  So, if the file you just created that you want to move is in the root of your vault folder, then the above code works fine.  However, if the file is not at the root location in your vault folder, it doesn't work.  

Here's some code I embedded in my templater script:
```javascript
console.log(`tp.file.folder(true) = ${tp.file.folder(true)}`)
console.log(`yearLink = ${yearLink}`)
const path = tp.file.folder(true) + "/" + yearLink;
console.log(`path = ${path}`)
console.log(`tp.file.exists(path) = ${tp.file.exists(path)}`)
console.log(`tp.file.exists(path + "/") = ${tp.file.exists(path + "/")}`) // to see if adding the slash made any difference -- it didn't
console.log(`tp.file.exists(yearlink) = ${tp.file.exists(yearLink)}`)
```
and here's the output from the console window (note that the full path in path **does** exist; also, I do have a sub-folder called "Obsidian" at the root of my vault -- it's not the name of vault folder):
```
tp.file.folder(true) = Obsidian/Journal/daily
yearLink = 2022
path = Obsidian/Journal/daily/2022   // this directory does exist
tp.file.exists(path) = false
tp.file.exists(path + "/") = false
tp.file.exists(yearlink) = true  // only the relative portion of the path is included
```
What makes this especially confusing is that app.vault.createFolder(dir) appears to want a path with folders relative to the vault folder (it also creates multiple sub-folders if it needs to make sure the path fully exists).  However, if all of the sub-folders currently exist, then the script exits with an error that the folder already exists.  If I was trying to move to a folder that had multiple path subfolders that didn't exist, testing for existence would be "problematic".  It seems like app.vault.createFolder should simply act as a noop if all the folders exist or tp.file.exists needs some enhancement.

It would be nice if tpfile.exists had a "relative" parameter (default false; like tp.file.folder) that allowed you to say you were specifying the entire path relative to the vault as opposed to relative to the current folder.

Let me know if I'm just missing something here.

**Reply by Fenrihr** [reply]
`tp.file.move` create the folder is not exist ... create all the path

**Comment by Juilio** [comment]
Hello, i tried to implement the code into my template, but if I use the template nothing happens. The Task is to create a new folder with the last Month as name, if an new month starts.
Following is my code:
![grafik](https://github.com/SilentVoid13/Templater/assets/143953362/24b5940e-feaa-423d-ae18-2d97ee75f071)

Any ideas why?

**Reply by Zachatoo** [reply]
Should be `tp.date.now("DD")` and `tp.date.yesterday("DD")` (notice the `"`). Otherwise it's trying to refer to a variable called `DD`.
